File: backend/app/src/main/kotlin/net/animeta/backend/controller/v2/WorkByTitleController.kt
package net.animeta.backend.controller.v2

import net.animeta.backend.dto.WorkDTO
import net.animeta.backend.exception.ApiException
import net.animeta.backend.model.User
import net.animeta.backend.repository.TitleMappingRepository
import net.animeta.backend.security.CurrentUser
import net.animeta.backend.serializer.WorkSerializer
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/v2/works/by-title")
class WorkByTitleController(val titleMappingRepository: TitleMappingRepository, val workSerializer: WorkSerializer) {
    @GetMapping
    fun get(@RequestParam title: String, @CurrentUser(required = false) currentUser: User?): WorkDTO {
        val work = titleMappingRepository.findOneByTitle(title)?.work ?: throw ApiException.notFound()
        return workSerializer.serialize(work, currentUser, full = true)
    }
}

File: backend-ts/src/controllers/work_by_title.controller.ts
import { Controller, Get, Query } from "@nestjs/common";
import { CurrentUser } from "src/auth/decorators";
import { User } from "src/entities/user.entity";
import { ApiException } from "./exceptions";
import { WorkService } from "src/services/work.service";
import { WorkSerializer } from "src/serializers/work.serializer";
import { WorkDTO } from "shared/types";
import { Repository } from "typeorm";
import { InjectRepository } from "@nestjs/typeorm";
import { TitleMapping } from "src/entities/title_mapping.entity";

@Controller('/api/v4/works/by-title')
export class WorkByTitleController {
  constructor(
    private workService: WorkService,
    @InjectRepository(TitleMapping) private titleMappingRepository: Repository<TitleMapping>,
    private workSerializer: WorkSerializer,
  ) {}

  @Get()
  async get(
    @Query('title') title: string,
    @CurrentUser({ required: false }) currentUser: User | null,
  ): Promise<WorkDTO> {
    const titleMapping = await this.titleMappingRepository.findOne({ where: {title} })
    if (!titleMapping) throw ApiException.notFound()
    const work = await this.workService.get(titleMapping.work_id)
    if (!work) throw ApiException.notFound()
    return this.workSerializer.serialize(work, currentUser, true)
  }
}
